File: src/main/java/backend/model/user/UserCore.java
package backend.model.user;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.hibernate.validator.constraints.Email;
import org.hibernate.validator.constraints.NotEmpty;

import javax.persistence.*;
import java.util.HashMap;
import java.util.Map;

@Entity
public class UserCore implements User {

    @Id
    @GeneratedValue
    private Integer id;

    @NotEmpty
    private String firstname;

    @NotEmpty
    private String lastname;

    @Email
    @Column(unique = true, nullable = false)
    private String email;

    // TODO: Not sure how to annotate this one
    private boolean isBlocked = false;

    @NotEmpty
    private String password;

    @NotEmpty
    private String gender;

    /*
     * cascade all operations to children
     * orphanRemoval = true allows removing a role from the database
     * if it gets removed from Map userRoles and the core is saved!
     * See: http://stackoverflow.com/a/2011546
     */
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER, orphanRemoval = true)
    private Map<Class<? extends UserRole>, UserRole> userRoles = new HashMap<>();

    @Override
    public String getFirstname() {
        return firstname;
    }

    @Override
    public void setFirstname(String firstname) {
        this.firstname = firstname;
    }

    @Override
    public String getLastname() {
        return lastname;
    }

    @Override
    public void setLastname(String lastname) {
        this.lastname = lastname;
    }

    @Override
    public String getEmail() {
        return email;
    }

    @Override
    public void setEmail(String email) {
        this.email = email;
    }

    @Override
    public boolean isBlocked() {
        return isBlocked;
    }

    public void setBlocked(boolean blocked) {
        isBlocked = blocked;
    }

    @Override
    public void setIsBlocked(boolean isBlocked) {
        this.isBlocked = isBlocked;
    }

    @Override
    public String getPassword() {
        return password;
    }

    @Override
    public void setPassword(String password) {
        this.password = password;
    }

    @Override
    public String getGender() {
        return gender;
    }

    @Override
    public void setGender(String gender) {
        this.gender = gender;
    }

    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public Map<Class<? extends UserRole>, UserRole> getUserRoles() {
        return userRoles;
    }

    public void setUserRoles(Map<Class<? extends UserRole>, UserRole> userRoles) {
        this.userRoles = userRoles;
    }

    @Override
    public UserRole addRole(Class<? extends UserRole> clazz) throws Exception {

        UserRole role;

        if (userRoles.containsKey(clazz)) {
            return userRoles.get(clazz);
        } else {
            role = UserRole.createFor(clazz, this);
            userRoles.put(clazz, role);
            return role;
        }
    }

    @Override
    public UserRole getRole(Class<? extends UserRole> clazz) {
        return userRoles.get(clazz);
    }

    @Override
    public boolean hasRole(Class<? extends UserRole> clazz) {
        return userRoles.containsKey(clazz);
    }

    @Override
    public UserRole removeRole(Class<? extends UserRole> clazz) {
        if (userRoles.containsKey(clazz)) {
            return userRoles.remove(clazz);
        } else {
            return null;
        }
    }

    @Override
    @JsonIgnore
    public UserCore getCore() {
        return this;
    }
}

File: src/main/java/backend/model/user/UserCore.kt
package backend.model.user

import backend.exceptions.DomainException
import backend.model.BasicEntity
import backend.model.media.Media
import backend.model.messaging.GroupMessage
import backend.model.messaging.Message
import backend.model.payment.Payment
import backend.model.posting.Comment
import backend.model.posting.Posting
import com.fasterxml.jackson.annotation.JsonIgnore
import org.hibernate.validator.constraints.Email
import org.hibernate.validator.constraints.NotEmpty
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import java.util.*
import javax.persistence.*
import kotlin.reflect.KClass

@Entity
open class UserCore : BasicEntity, User {

    constructor() : super() {
        this.isBlocked = true
        this.profilePic = Media("image")
    }

    @Email
    @Column(unique = true, nullable = false)
    override lateinit var email: String

    @NotEmpty
    override lateinit var passwordHash: String

    override var firstname: String? = null

    override var lastname: String? = null

    override var isBlocked = false

    override var gender: String? = null

    @OneToOne(cascade = arrayOf(CascadeType.ALL), orphanRemoval = true)
    override var profilePic: Media

    @OneToMany(mappedBy = "user", cascade = arrayOf(CascadeType.ALL))
    val postings: MutableList<Posting> = ArrayList()

    @OneToMany(mappedBy = "user", cascade = arrayOf(CascadeType.ALL))
    val comments: MutableList<Comment> = ArrayList()

    @ManyToMany(mappedBy = "users", cascade = arrayOf(CascadeType.ALL))
    val groupMessages: MutableList<GroupMessage> = ArrayList()

    @OneToMany(mappedBy = "creator", cascade = arrayOf(CascadeType.ALL))
    val messages: MutableList<Message> = ArrayList()

    @OneToMany(mappedBy = "user", cascade = arrayOf(CascadeType.ALL))
    val payments: MutableList<Payment> = ArrayList()

    /*
     * cascade all operations to children
     * orphanRemoval = true allows removing a role from the database
     * if it gets removed from Map userRoles and the core is saved!
     * See: http://stackoverflow.com/a/2011546
     */
    @OneToMany(cascade = arrayOf(CascadeType.ALL), fetch = FetchType.EAGER, orphanRemoval = true)
    private var userRoles: MutableMap<Class<out UserRole>, UserRole> = HashMap()

    private var activationToken: String? = null

    fun getAuthorities(): Collection<GrantedAuthority> {
        return this.userRoles.values.map { BasicGrantedAuthority(it) }
    }

    override fun activate(token: String) {
        if (isActivationTokenCorrect(token)) {
            this.isBlocked = false
            this.activationToken = null;
        } else {
            throw DomainException("Provided token $token does not match the activation token")
        }
    }

    override fun isActivationTokenCorrect(token: String): Boolean {
        return this.activationToken == token
    }

    override fun createActivationToken(): String {
        this.activationToken = UUID.randomUUID().toString()
        return this.activationToken!!
    }

    override fun isActivated(): Boolean {
        return !isBlocked;
    }

    override fun setNewPassword(password: String, token: String) {

        //successful password reset is also email validation
        this.activate(token)
        this.passwordHash = BCryptPasswordEncoder().encode(password)
    }

    // This cast will always succeed because the specific type of the value / object
    // in the Map<Class<out UserRole>, UserRole> always matches the one defined with T
    // This is guaranteed as the key of the map always expresses the exact class of
    // the object stored as the value
    @Suppress("UNCHECKED_CAST")
    override fun <T : UserRole> getRole(clazz: KClass<T>): T? = userRoles[clazz.java] as? T?

    override fun <T : UserRole> hasRole(clazz: KClass<T>): Boolean = userRoles.containsKey(clazz.java)

    @Suppress("UNCHECKED_CAST")
    override fun <T : UserRole> removeRole(clazz: KClass<T>): T? = userRoles.remove(clazz.java) as? T?

    @Throws(Exception::class)
    @Suppress("UNCHECKED_CAST")
    override fun <T : UserRole> addRole(clazz: KClass<T>): T {
        val role: UserRole
        if (userRoles.containsKey(clazz.java)) throw DomainException("User already has role $clazz")
        role = UserRole.createFor(clazz.java, this)
        userRoles.put(clazz.java, role)
        return role as T
    }

    override val core: UserCore
        @JsonIgnore
        get() = this

    @PreRemove
    fun preRemove() {
        this.payments.forEach { it.user = null }
        this.payments.clear()
    }
}

class BasicGrantedAuthority : GrantedAuthority {

    private val internal: String

    constructor(userRole: UserRole) {
        this.internal = userRole.authority
    }

    override fun getAuthority(): String? {
        return internal
    }
}
